File: python/plugin_paths.py
import json
import os
import stat
import tempfile
from contextlib import contextmanager


CONFIG_SCHEMA = 1


def config_path():
    configured = os.environ.get("CASCADE_CONFIG_FILE")
    if configured:
        return os.path.abspath(os.path.expanduser(configured))
    config_home = os.environ.get("XDG_CONFIG_HOME")
    if not config_home:
        home = os.environ.get("HOME") or os.environ.get("USERPROFILE") or os.path.expanduser("~")
        config_home = os.path.join(home, ".config")
    return os.path.join(os.path.abspath(os.path.expanduser(config_home)), "cascade", "config.json")


def canonical_prefix(path):
    return os.path.realpath(os.path.abspath(os.path.expanduser(path)))


def _empty_config():
    return {"schema": CONFIG_SCHEMA, "plugin_prefixes": []}


def load_config(path=None):
    path = path or config_path()
    if not os.path.exists(path):
        return _empty_config()
    with open(path, "r", encoding="utf-8") as source:
        document = json.load(source)
    if not isinstance(document, dict):
        raise RuntimeError(f"Cascade config must be a JSON object: {path}")
    if document.get("schema") != CONFIG_SCHEMA:
        raise RuntimeError(f"Unsupported Cascade config schema in {path}")
    entries = document.get("plugin_prefixes", [])
    if not isinstance(entries, list):
        raise RuntimeError(f"Cascade config plugin_prefixes must be a list: {path}")

    normalized = []
    seen = set()
    for entry in entries:
        if isinstance(entry, str):
            raw_path = entry
            enabled = True
        elif isinstance(entry, dict):
            raw_path = entry.get("path")
            enabled = entry.get("enabled", True)
        else:
            raise RuntimeError(f"Invalid plugin prefix entry in {path}: {entry!r}")
        if not isinstance(raw_path, str) or not raw_path.strip():
            raise RuntimeError(f"Plugin prefix path must be a non-empty string in {path}")
        if not isinstance(enabled, bool):
            raise RuntimeError(f"Plugin prefix enabled flag must be boolean in {path}")
        prefix = canonical_prefix(raw_path)
        if prefix in seen:
            continue
        seen.add(prefix)
        normalized.append({"path": prefix, "enabled": enabled})
    return {"schema": CONFIG_SCHEMA, "plugin_prefixes": normalized}


def configured_plugin_prefixes(path=None):
    return [
        entry["path"]
        for entry in load_config(path).get("plugin_prefixes", [])
        if entry.get("enabled", True)
    ]


def _atomic_write_config(document, path):
    directory = os.path.dirname(path)
    _ensure_real_directory(directory)
    descriptor, temporary = tempfile.mkstemp(prefix="config.", suffix=".tmp", dir=directory)
    try:
        with os.fdopen(descriptor, "w", encoding="utf-8") as output:
            json.dump(document, output, indent=2, sort_keys=True)
            output.write("\n")
            output.flush()
            os.fsync(output.fileno())
        os.chmod(temporary, 0o600)
        if os.rename in os.supports_dir_fd:
            directory_descriptor = _open_real_directory(directory)
            try:
                os.replace(
                    os.path.basename(temporary),
                    os.path.basename(path),
                    src_dir_fd=directory_descriptor,
                    dst_dir_fd=directory_descriptor,
                )
                os.fsync(directory_descriptor)
            finally:
                os.close(directory_descriptor)
        else:
            os.replace(temporary, path)
    finally:
        if os.path.exists(temporary):
            os.remove(temporary)


@contextmanager
def _config_lock(path):
    directory = os.path.dirname(path)
    _ensure_real_directory(directory)
    lock_path = path + ".lock"
    flags = os.O_RDWR | os.O_CREAT
    if hasattr(os, "O_BINARY"):
        flags |= os.O_BINARY
    if hasattr(os, "O_NOFOLLOW"):
        flags |= os.O_NOFOLLOW
    if os.open in os.supports_dir_fd:
        directory_descriptor = _open_real_directory(directory)
        try:
            descriptor = os.open(os.path.basename(lock_path), flags, 0o600, dir_fd=directory_descriptor)
        finally:
            os.close(directory_descriptor)
    else:
        descriptor = os.open(lock_path, flags, 0o600)
    if not stat.S_ISREG(os.fstat(descriptor).st_mode):
        os.close(descriptor)
        raise RuntimeError(f"Cascade config lock must be a regular file: {lock_path}")
    with os.fdopen(descriptor, "a+b") as lock:
        if os.name == "nt":
            import msvcrt
            lock.seek(0)
            if os.path.getsize(lock_path) == 0:
                lock.write(b"\0")
                lock.flush()
            lock.seek(0)
            msvcrt.locking(lock.fileno(), msvcrt.LK_LOCK, 1)
            try:
                yield
            finally:
                lock.seek(0)
                msvcrt.locking(lock.fileno(), msvcrt.LK_UNLCK, 1)
        else:
            import fcntl
            fcntl.flock(lock.fileno(), fcntl.LOCK_EX)
            try:
                yield
            finally:
                fcntl.flock(lock.fileno(), fcntl.LOCK_UN)


def _ensure_real_directory(directory):
    directory = os.path.abspath(directory)
    missing = []
    current = directory
    while not os.path.lexists(current):
        missing.append(current)
        parent = os.path.dirname(current)
        if parent == current:
            break
        current = parent
    while True:
        metadata = os.lstat(current)
        if stat.S_ISLNK(metadata.st_mode):
            raise RuntimeError(f"Cascade config parent must not be a symbolic link: {current}")
        if not stat.S_ISDIR(metadata.st_mode):
            raise RuntimeError(f"Cascade config parent is not a directory: {current}")
        parent = os.path.dirname(current)
        if parent == current:
            break
        current = parent
    for path in reversed(missing):
        try:
            os.mkdir(path, 0o700)
        except FileExistsError:
            metadata = os.lstat(path)
            if stat.S_ISLNK(metadata.st_mode) or not stat.S_ISDIR(metadata.st_mode):
                raise RuntimeError(f"Cascade config parent is not a real directory: {path}")


def _open_real_directory(directory):
    flags = os.O_RDONLY | getattr(os, "O_DIRECTORY", 0) | getattr(os, "O_NOFOLLOW", 0)
    descriptor = os.open(os.sep, flags)
    try:
        for component in os.path.abspath(directory).split(os.sep):
            if not component:
                continue
            next_descriptor = os.open(component, flags, dir_fd=descriptor)
            os.close(descriptor)
            descriptor = next_descriptor
        result = descriptor
        descriptor = -1
        return result
    finally:
        if descriptor >= 0:
            os.close(descriptor)


def add_plugin_prefix(prefix, path=None):
    path = path or config_path()
    prefix = canonical_prefix(prefix)
    with _config_lock(path):
        document = load_config(path)
        entries = document["plugin_prefixes"]
        for entry in entries:
            if entry["path"] == prefix:
                changed = not entry.get("enabled", True)
                entry["enabled"] = True
                if changed:
                    _atomic_write_config(document, path)
                return changed
        entries.append({"path": prefix, "enabled": True})
        _atomic_write_config(document, path)
    return True


def remove_plugin_prefix(prefix, path=None):
    path = path or config_path()
    prefix = canonical_prefix(prefix)
    with _config_lock(path):
        document = load_config(path)
        original = list(document["plugin_prefixes"])
        document["plugin_prefixes"] = [entry for entry in original if entry["path"] != prefix]
        if len(document["plugin_prefixes"]) == len(original):
            return False
        _atomic_write_config(document, path)
    return True


def plugin_layout(prefix):
    prefix = canonical_prefix(prefix)
    return {
        "prefix": prefix,
        "cpp": os.path.join(prefix, "lib", "cascade", "plugin"),
        "python": os.path.join(prefix, "lib", "cascade", "pyplugin"),
        "include": os.path.join(prefix, "include", "cascade", "plugin"),
        "trust_store": os.path.join(prefix, "share", "cascade", "trusted_keys"),
    }


def unique_paths(paths):
    result = []
    seen = set()
    for path in paths:
        canonical = os.path.realpath(os.path.abspath(os.path.expanduser(path)))
        if canonical in seen:
            continue
        seen.add(canonical)
        result.append(canonical)
    return result

File: src/PluginPaths.cc
#include "PluginPaths.hh"

#include <dlfcn.h>
#include <algorithm>
#include <cstdlib>
#include <cerrno>
#include <cstring>
#include <filesystem>
#include <fstream>
#include <nlohmann/json.hpp>
#include <fcntl.h>
#include <set>
#include <stdexcept>
#include <sys/file.h>
#include <sys/stat.h>
#include <unistd.h>

namespace
{
namespace fs = std::filesystem;

fs::path Expanded(const std::string &value)
{
    if (value == "~" || value.rfind("~/", 0) == 0)
    {
        const char *home = std::getenv("HOME");
        if (home && *home) return fs::path(home) / value.substr(value == "~" ? 1 : 2);
    }
    return value;
}

std::string Absolute(const fs::path &path)
{
    std::error_code error;
    fs::path result = fs::weakly_canonical(fs::absolute(path), error);
    if (error) result = fs::absolute(path).lexically_normal();
    return result.string();
}

std::string LexicalAbsolute(const fs::path &path) { return fs::absolute(path).lexically_normal().string(); }

void EnsureRealDirectory(const fs::path &directory)
{
    const fs::path absolute = fs::absolute(directory).lexically_normal();
    fs::path current = absolute.root_path();
    for (const auto &component : absolute.relative_path())
    {
        current /= component;
        std::error_code error;
        const auto status = fs::symlink_status(current, error);
        if (error == std::errc::no_such_file_or_directory || status.type() == fs::file_type::not_found)
        {
            error.clear();
            if (!fs::create_directory(current, error) && error)
                throw std::runtime_error("Cannot create Cascade config directory: " + current.string());
            fs::permissions(current, fs::perms::owner_all, fs::perm_options::replace, error);
            continue;
        }
        if (error) throw std::runtime_error("Cannot inspect Cascade config parent: " + current.string());
        if (fs::is_symlink(status))
            throw std::runtime_error("Cascade config parent must not be a symbolic link: " + current.string());
        if (!fs::is_directory(status))
            throw std::runtime_error("Cascade config parent is not a directory: " + current.string());
    }
}

class ConfigLock
{
  public:
    explicit ConfigLock(const fs::path &configPath)
    {
        EnsureRealDirectory(configPath.parent_path());
        const std::string lockPath = configPath.string() + ".lock";
        m_Descriptor = open(lockPath.c_str(), O_RDWR | O_CREAT | O_NOFOLLOW | O_CLOEXEC, 0600);
        if (m_Descriptor < 0) throw std::runtime_error("Cannot open Cascade config lock: " + lockPath);
        struct stat metadata{};
        if (fstat(m_Descriptor, &metadata) != 0 || !S_ISREG(metadata.st_mode))
        {
            close(m_Descriptor);
            throw std::runtime_error("Cascade config lock must be a regular file: " + lockPath);
        }
        if (flock(m_Descriptor, LOCK_EX) != 0)
        {
            close(m_Descriptor);
            throw std::runtime_error("Cannot lock Cascade config: " + lockPath);
        }
    }

    ~ConfigLock()
    {
        if (m_Descriptor >= 0)
        {
            flock(m_Descriptor, LOCK_UN);
            close(m_Descriptor);
        }
    }

    ConfigLock(const ConfigLock &) = delete;
    ConfigLock &operator=(const ConfigLock &) = delete;

  private:
    int m_Descriptor = -1;
};

nlohmann::json ConfigJson(const PluginConfig &config)
{
    nlohmann::json entries = nlohmann::json::array();
    for (const auto &entry : config.Prefixes) entries.push_back({{"enabled", entry.Enabled}, {"path", entry.Path}});
    return {{"plugin_prefixes", std::move(entries)}, {"schema", config.Schema}};
}

void AtomicWriteConfig(const PluginConfig &config, const fs::path &configPath)
{
    EnsureRealDirectory(configPath.parent_path());
    std::string temporary = (configPath.parent_path() / "config.XXXXXX").string();
    std::vector<char> name(temporary.begin(), temporary.end());
    name.push_back('\0');
    const int descriptor = mkstemp(name.data());
    if (descriptor < 0) throw std::runtime_error("Cannot create temporary Cascade config");
    const std::string payload = ConfigJson(config).dump(2) + '\n';
    bool success = false;
    try
    {
        std::size_t offset = 0;
        while (offset < payload.size())
        {
            const ssize_t written = write(descriptor, payload.data() + offset, payload.size() - offset);
            if (written < 0)
            {
                if (errno == EINTR) continue;
                throw std::runtime_error("Cannot write temporary Cascade config");
            }
            offset += static_cast<std::size_t>(written);
        }
        if (fsync(descriptor) != 0 || fchmod(descriptor, 0600) != 0)
            throw std::runtime_error("Cannot flush temporary Cascade config");
        if (close(descriptor) != 0) throw std::runtime_error("Cannot close temporary Cascade config");
        if (rename(name.data(), configPath.c_str()) != 0)
            throw std::runtime_error("Cannot replace Cascade config: " + configPath.string());
        const int directory = open(configPath.parent_path().c_str(), O_RDONLY | O_DIRECTORY | O_CLOEXEC);
        if (directory >= 0)
        {
            fsync(directory);
            close(directory);
        }
        success = true;
    }
    catch (...)
    {
        if (!success) close(descriptor);
        unlink(name.data());
        throw;
    }
}
} // namespace

std::string PluginPaths::ConfigPath()
{
    if (const char *configured = std::getenv("CASCADE_CONFIG_FILE"); configured && *configured)
        return LexicalAbsolute(Expanded(configured));
    if (const char *configHome = std::getenv("XDG_CONFIG_HOME"); configHome && *configHome)
        return LexicalAbsolute(Expanded(configHome) / "cascade" / "config.json");
    const char *home = std::getenv("HOME");
    if (!home || !*home) home = std::getenv("USERPROFILE");
    return LexicalAbsolute((home && *home ? fs::path(home) : fs::path(".")) / ".config" / "cascade" / "config.json");
}

std::string PluginPaths::RuntimePrefix()
{
    if (const char *configured = std::getenv("CASCADE_PREFIX"); configured && *configured)
        return Absolute(Expanded(configured));
    Dl_info libraryInfo{};
    if (dladdr(reinterpret_cast<void *>(&PluginPaths::RuntimePrefix), &libraryInfo) != 0 && libraryInfo.dli_fname)
        return Absolute(fs::path(libraryInfo.dli_fname).parent_path().parent_path());
    const char *home = std::getenv("HOME");
    return Absolute(home && *home ? fs::path(home) / ".local" : fs::path("."));
}

std::string PluginPaths::CanonicalPrefix(const std::string &path) { return Absolute(Expanded(path)); }

std::vector<std::string> PluginPaths::Unique(const std::vector<std::string> &paths)
{
    std::vector<std::string> result;
    std::set<std::string> seen;
    for (const auto &path : paths)
    {
        const std::string canonical = CanonicalPrefix(path);
        if (seen.insert(canonical).second) result.push_back(canonical);
    }
    return result;
}

PluginConfig PluginPaths::LoadConfig(const std::string &path)
{
    const fs::path configPath = path.empty() ? ConfigPath() : LexicalAbsolute(Expanded(path));
    if (!fs::exists(configPath)) return {};
    std::ifstream input(configPath);
    if (!input) throw std::runtime_error("Cannot open Cascade config: " + configPath.string());
    nlohmann::json config;
    input >> config;
    if (!config.is_object() || config.value("schema", 0) != 1)
        throw std::runtime_error("Unsupported Cascade config schema: " + configPath.string());
    const auto entries = config.value("plugin_prefixes", nlohmann::json::array());
    if (!entries.is_array()) throw std::runtime_error("Cascade config plugin_prefixes must be a list");
    PluginConfig result;
    std::set<std::string> seen;
    for (const auto &entry : entries)
    {
        std::string value;
        bool enabled = true;
        if (entry.is_string())
            value = entry.get<std::string>();
        else if (entry.is_object())
        {
            if (!entry.contains("path") || !entry.at("path").is_string())
                throw std::runtime_error("Plugin prefix path must be a non-empty string");
            value = entry.at("path").get<std::string>();
            if (entry.contains("enabled") && !entry.at("enabled").is_boolean())
                throw std::runtime_error("Plugin prefix enabled flag must be boolean");
            enabled = entry.value("enabled", true);
        }
        else
            throw std::runtime_error("Invalid plugin prefix entry");
        if (value.empty()) throw std::runtime_error("Plugin prefix path must be a non-empty string");
        value = CanonicalPrefix(value);
        if (seen.insert(value).second) result.Prefixes.push_back({std::move(value), enabled});
    }
    return result;
}

std::vector<std::string> PluginPaths::ConfiguredPrefixes(const std::string &path)
{
    std::vector<std::string> result;
    for (const auto &entry : LoadConfig(path).Prefixes)
        if (entry.Enabled) result.push_back(entry.Path);
    return result;
}

bool PluginPaths::AddPrefix(const std::string &prefixValue, const std::string &path)
{
    const fs::path configPath = path.empty() ? ConfigPath() : LexicalAbsolute(Expanded(path));
    const std::string prefix = CanonicalPrefix(prefixValue);
    ConfigLock lock(configPath);
    auto config = LoadConfig(configPath.string());
    for (auto &entry : config.Prefixes)
    {
        if (entry.Path != prefix) continue;
        if (entry.Enabled) return false;
        entry.Enabled = true;
        AtomicWriteConfig(config, configPath);
        return true;
    }
    config.Prefixes.push_back({prefix, true});
    AtomicWriteConfig(config, configPath);
    return true;
}

bool PluginPaths::RemovePrefix(const std::string &prefixValue, const std::string &path)
{
    const fs::path configPath = path.empty() ? ConfigPath() : LexicalAbsolute(Expanded(path));
    const std::string prefix = CanonicalPrefix(prefixValue);
    ConfigLock lock(configPath);
    auto config = LoadConfig(configPath.string());
    const auto original = config.Prefixes.size();
    config.Prefixes.erase(std::remove_if(config.Prefixes.begin(), config.Prefixes.end(),
                                         [&](const PluginPrefix &entry) { return entry.Path == prefix; }),
                          config.Prefixes.end());
    if (config.Prefixes.size() == original) return false;
    AtomicWriteConfig(config, configPath);
    return true;
}

PluginLayout PluginPaths::Layout(const std::string &prefixValue)
{
    PluginLayout layout;
    layout.Prefix = CanonicalPrefix(prefixValue);
    const fs::path prefix(layout.Prefix);
    layout.Cpp = (prefix / "lib" / "cascade" / "plugin").string();
    layout.Python = (prefix / "lib" / "cascade" / "pyplugin").string();
    layout.Include = (prefix / "include" / "cascade" / "plugin").string();
    layout.TrustStore = (prefix / "share" / "cascade" / "trusted_keys").string();
    return layout;
}

std::vector<PluginLayout> PluginPaths::RuntimeLayouts()
{
    std::vector<PluginLayout> candidates;
    const char *trust = std::getenv("CASCADE_PLUGIN_TRUST_STORE");
    const char *cpp = std::getenv("CASCADE_PLUGIN_DIR");
    const char *python = std::getenv("CASCADE_PYPLUGIN_DIR");
    if ((cpp && *cpp) || (python && *python))
    {
        PluginLayout environment = Layout(RuntimePrefix());
        environment.Prefix = "environment";
        environment.Source = "environment";
        environment.Cpp = cpp && *cpp ? CanonicalPrefix(cpp) : "";
        environment.Python = python && *python ? CanonicalPrefix(python) : "";
        if (trust && *trust) environment.TrustStore = CanonicalPrefix(trust);
        candidates.push_back(std::move(environment));
    }
    for (const auto &prefix : ConfiguredPrefixes())
    {
        auto layout = Layout(prefix);
        layout.Source = "config";
        if (trust && *trust) layout.TrustStore = CanonicalPrefix(trust);
        candidates.push_back(std::move(layout));
    }
    auto runtime = Layout(RuntimePrefix());
    runtime.Source = "cascade";
    if (trust && *trust) runtime.TrustStore = CanonicalPrefix(trust);
    candidates.push_back(std::move(runtime));

    std::vector<PluginLayout> result;
    std::set<std::string> seen;
    for (auto &layout : candidates)
    {
        const std::string key = layout.Cpp + '\n' + layout.Python + '\n' + layout.TrustStore;
        if (seen.insert(key).second) result.push_back(std::move(layout));
    }
    return result;
}

std::vector<std::string> PluginPaths::Roots(const std::string &language)
{
    std::vector<std::string> roots;
    for (const auto &layout : RuntimeLayouts()) roots.push_back(language == "python" ? layout.Python : layout.Cpp);
    return Unique(roots);
}

std::string PluginPaths::TrustStoreForRoot(const std::string &pluginRoot)
{
    if (const char *configured = std::getenv("CASCADE_PLUGIN_TRUST_STORE"); configured && *configured)
        return CanonicalPrefix(configured);
    fs::path prefix = fs::path(pluginRoot);
    for (int level = 0; level < 3 && prefix.has_parent_path(); ++level) prefix = prefix.parent_path();
    return Absolute(prefix / "share" / "cascade" / "trusted_keys");
}
